tools/autograd/gen_variable_type.py

```diff
@@ -117,6 +139,58 @@
 CALL_VIA_DERIVED = CodeTemplate("""\
 baseType->${method_prefix_derived}${base_name}(${unpacked_args})""")
 
+SAVE_TENSOR_STORAGE = CodeTemplate("""\
+Storage ${tensor_name}_storage_saved =
+  (${tensor_name}.defined() && !${tensor_name}.is_sparse()) ? ${tensor_name}.storage() : Storage();
+""")
+
+ENFORCE_SAME_TENSOR_STORAGE = CodeTemplate("""\
+if (${tensor_name}_storage_saved) AT_ASSERT(${tensor_name}_storage_saved.is_alias_of(${tensor_name}.storage()));
+""")
+
+SAVE_TENSORLIST_STORAGE = CodeTemplate("""\
+std::vector<Storage> ${tensorlist_name}_storage_saved(${tensorlist_name}.size());
+for (size_t i=0; i<${tensorlist_name}.size(); i++) {
+  ${tensorlist_name}_storage_saved[i] =
+    (${tensorlist_name}[i].defined() && !${tensorlist_name}[i].is_sparse()) ?
+      ${tensorlist_name}[i].storage() : Storage();
+}
+""")
+
+ENFORCE_SAME_TENSORLIST_STORAGE = CodeTemplate("""\
+for (size_t i=0; i<${tensorlist_name}.size(); i++) {
+  if (${tensorlist_name}_storage_saved[i]) {
+    AT_ASSERT(${tensorlist_name}_storage_saved[i].is_alias_of(${tensorlist_name}[i].storage()));
+  }
+}
+""")
+
+SAVE_TENSOR_IMPL = CodeTemplate("""\
+c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> ${tensor_name}_impl_saved;
+if (${tensor_name}.defined() && !${tensor_name}.is_sparse())
```

Comment:
do sparse tensors not have intrusive ptrs?

Reply:
Sparse tensors' intrusive ptrs are in `indices` and `values`, however the function for setting `indices` and `values` of a sparse tensor (`set_indices_and_values_unsafe`) always replace them with new tensors, and not preserving their original TensorImpl and Storage. This is expected because the number of elements in sparse tensor is dynamic, and we can't reliably preserve the same Storage pointer in cases where we need to add more elements to the tensor (thus needing to expand storage in `indices` and `values`).

Reply:
fixed